File: src/leetcode/binary_search/162_find_peak_element.rb
# frozen_string_literal: true

# LeetCode 75
# https://leetcode.com/problems/find-peak-element/?envType=study-plan-v2&envId=leetcode-75

# Time complexity: O(log n)
# Space complexity: O(1)

# @param {Integer[]} nums
# @return {Integer}
def find_peak_element(nums)
  left = 0
  right = nums.length - 1

  while left < right
    mid = left + ((right - left) / 2)

    num = nums[mid]
    right_num = nums[mid + 1]

    if num > right_num
      right = mid
    else
      left = mid + 1
    end
  end

  left # at the end of while, left == right
end

File: src/leetcode/binary_search/162_find_peak_element.ts
// LeetCode 75
// https://leetcode.com/problems/find-peak-element/?envType=study-plan-v2&envId=leetcode-75

// Time complexity: O(log n)
// Space complexity: O(1)

const findPeakElement = (nums: number[]): number => {
  let left = 0;
  let right = nums.length;

  while (left < right) {
    const mid = Math.floor(left + (right - left) / 2);

    const num = nums[mid];
    const rightNum = nums[mid + 1];

    if (num > rightNum) {
      right = mid;
    } else {
      left = mid + 1;
    }
  }

  return left;
};

export default findPeakElement;
